File: PlaylistApp.Server/Services/GoalLikeServices/GoalLikeService.cs
using Microsoft.EntityFrameworkCore;
using PlaylistApp.Server.Data;
using PlaylistApp.Server.DTOs;
using PlaylistApp.Server.Requests.AddRequests;
using PlaylistApp.Server.Requests.DeleteRequests;
using PlaylistApp.Server.Requests.GetRequests;
using PlaylistApp.Server.Requests.UpdateRequests;

namespace PlaylistApp.Server.Services.GoalLikeServices;

public class GoalLikeService : IGoalLikeService
{
    private readonly IDbContextFactory<PlaylistDbContext> dbContextFactory;

    public GoalLikeService(IDbContextFactory<PlaylistDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }

    public async Task<bool> AddGoalLike(AddGoalLikeRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var user = await context.UserAccounts
            .Where(x => x.Guid == request.UserId)
            .FirstOrDefaultAsync();

        if (user == null)
        {
            return false;
        }

        var possibleGoalLike = await context.GoalLikes
            .Where(x => x.User.Guid == request.UserId)
            .Where(x => x.GoalId == request.GoalId)
            .FirstOrDefaultAsync();

        if (possibleGoalLike != null)
        {
            return false;
        }

        GoalLike newGoalLike = new GoalLike()
        {
            DateLiked = DateTime.UtcNow,
            GoalId = request.GoalId,
            UserId = user.Id,
            IsLike = request.IsLike
        };

        await context.AddAsync(newGoalLike);
        await context.SaveChangesAsync();
        return true;
    }

    public async Task<GoalLikeDTO> GetGoalLike(GetGoalLikeRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var goalLike = await context.GoalLikes
            .Include(x => x.User)
                .ThenInclude(x => x.UserImage)
            .Include(x => x.Goal)
                .ThenInclude(x => x.Achievement)
                    .ThenInclude(x => x.PlatformGame)
                        .ThenInclude(x => x.Game)
            .Where(x => x.User.Guid == request.UserId)
            .Where(x => x.Id == request.GoalId)
            .FirstOrDefaultAsync();

        if (goalLike == null)
        {
            return new GoalLikeDTO();
        }

        return goalLike.ToDTO();
    }

    public async Task<List<GoalDTO>> GetGoalLikesFromUser(Guid userId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var goalLikes = await context.GoalLikes
            .Include(x => x.User)
                .ThenInclude(x => x.UserImage)
            .Include(x => x.Goal)
                .ThenInclude(x => x.Achievement)
                    .ThenInclude(x => x.PlatformGame)
                        .ThenInclude(x => x.Game)
            .Where(x => x.User.Guid == userId)
            .ToListAsync();

        return goalLikes.Select(x => x.Goal.ToDTO()).ToList();   
    }

    public async Task<bool> RemoveGoalLike(RemoveGoalLikeRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var goalLike = await context.GoalLikes
            .Where(x => x.User.Guid == request.UserId)
            .Where(x => x.GoalId == request.GoalId)
            .FirstOrDefaultAsync();

        if (goalLike == null)
        {
            return false;
        }

        context.Remove(goalLike);
        await context.SaveChangesAsync();
        return true;
    }

    public async Task<bool> UpdateGoalLike(UpdateGoalLikeRequest request)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var goalLikeUnderChange = await context.GoalLikes
            .Where(x => x.Id == request.Id)
            .FirstOrDefaultAsync();

        if (goalLikeUnderChange == null)
        {
            return false;
        }

        goalLikeUnderChange.IsLike = request.IsLike;
        goalLikeUnderChange.DateLiked = DateTime.UtcNow;

        context.Update(goalLikeUnderChange);
        await context.SaveChangesAsync();
        return true;
    }
}

File: playlistapp.client/src/ApiServices/GoalLikeService.ts
import { Goal } from "@/@types/goal";
import { GoalLike } from "@/@types/goalLike";
import { AddGoalLikeRequest } from "@/@types/Requests/AddRequests/addGoalLikeRequest";
import { RemoveGoalLikerequest } from "@/@types/Requests/DeleteRequests/removeGoalLikeRequest";
import { GetGoalLikeRequest } from "@/@types/Requests/GetRequests/getGoalLikeRequest";
import { UpdateGoalLikeRequest } from "@/@types/Requests/UpdateRequests/updateGoalLikeRequest";
import axios from "axios";
import { AuthenticationUtils } from "./AuthenticationUtils";

export const GoalLikeService = {
  addGoalLike: async (addGoalLikeRequest: AddGoalLikeRequest) => {
    if (!addGoalLikeRequest) {
      console.error("Add goal like request is undefined or empty");
      throw new Error("Add goal like request must be provided");
    }
    try {
      let jwtToken = AuthenticationUtils.GetJwtToken();
      const response = await axios.post<boolean>(
        `${import.meta.env.VITE_URL}/GoalLike/addgoallike`,
        addGoalLikeRequest,
        {
          headers: {
            Authorization: `Bearer ${jwtToken}`,
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to add goal like");
      throw error;
    }
  },
  getGoalLikesFromUser: async (userId: string) => {
    if (!userId) {
      console.error("User id is undefined or empty");
      throw new Error("User id must be provided");
    }
    try {
      const response = await axios.get<Goal[]>(
        `${import.meta.env.VITE_URL}/GoalLike/getgoallikesfromuser`,
        {
          params: {
            userId,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get goal likes from user");
      throw error;
    }
  },
  getGoalLike: async (getGoalLikeRequest: GetGoalLikeRequest) => {
    if (!getGoalLikeRequest) {
      console.error("Get goal like request is undefined or empty");
      throw new Error("Get goal like request must be provided");
    }
    try {
      const response = await axios.post<GoalLike>(
        `${import.meta.env.VITE_URL}/GoalLike/getgoallike`,
        getGoalLikeRequest,
        {
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get goal like from request");
      throw error;
    }
  },
  updateGoalLike: async (updateGoalLikeRequest: UpdateGoalLikeRequest) => {
    if (!updateGoalLikeRequest) {
      console.error("Update goal like request is undefined or empty");
      throw new Error("Update goal like request must be provided");
    }
    try {
      let jwtToken = AuthenticationUtils.GetJwtToken();
      const response = await axios.patch<boolean>(
        `${import.meta.env.VITE_URL}/GoalLike/updategoallike`,
        updateGoalLikeRequest,
        {
          headers: {
            Authorization: `Bearer ${jwtToken}`,
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to update goal like");
      throw error;
    }
  },
  removeGoalLike: async (removeGoalLikeRequest: RemoveGoalLikerequest) => {
    if (!removeGoalLikeRequest) {
      console.error("Remove goal like request is unefined or empty");
      throw new Error("Remove goal like request must be provided");
    }
    try {
      let jwtToken = AuthenticationUtils.GetJwtToken();
      const response = await axios.post<boolean>(
        `${import.meta.env.VITE_URL}/GoalLike/removegoallike`,
        removeGoalLikeRequest,
        {
          headers: {
            Authorization: `Bearer ${jwtToken}`,
            "Content-Type": "application/json",
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to remove goal like");
      throw error;
    }
  },
};
